# 4.5 Data Consistency

In [5]:
#Import Pandas Numpy and OS
import pandas as pd
import numpy as np
import os

In [6]:
#Import products data
df_prods = pd.read_csv('/Users/ariedlin/Desktop/Career Foundry Assignments/8-24 Instacart Blanket Analysis Real/2 Data/1 Original Data/products.csv')

In [7]:
#Import Orders data from Prepared Data folder
df_ords = pd.read_csv('/Users/ariedlin/Desktop/Career Foundry Assignments/8-24 Instacart Blanket Analysis Real/2 Data/2 Prepared Data/orders_wrangled.csv')

In [8]:
#Create a test dataframe
df_test = pd.DataFrame()

In [9]:
#Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [10]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [11]:
#Check for mixed type data
for col in df_test.columns:
    # Compare types in the column with the type of the first row
    weird = (df_test[col].apply(type) != type(df_test[col].iloc[0]))
    
    # If any row has a different type, print the column name
    if weird.any():
        print(col)

mix


In [12]:
#Change the data type for df_test to a string
df_test['mix'] = df_test['mix'].astype('str')

# Missing Data

In [14]:
#Check df_prods for missing data
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [15]:
#Create a subset for missing values in product_name
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [16]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [17]:
#Find number of rows in current df_prods dataframe
df_prods.shape

(49693, 5)

In [18]:
#Create new dataframe without the missing product_name rows
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [19]:
#Check number of rows in clean dataframe
df_prods_clean.shape

(49677, 5)

# Duplicate Data

In [21]:
#Create a subset for duplicate data
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [22]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [23]:
#Find current number of rows in df_prods_clean dataframe
df_prods_clean.shape

(49677, 5)

In [24]:
#Remove all duplicate values
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [25]:
#Check number of rows with duplicates removed
df_prods_clean_no_dups.shape

(49672, 5)

In [26]:
#Export df_prods_clean_no_dups as csv
df_prods_clean_no_dups.to_csv('/Users/ariedlin/Desktop/Career Foundry Assignments/8-24 Instacart Blanket Analysis Real/2 Data/2 Prepared Data/clean_data.csv')

# Exercise 4.5: Data Consistency Checks

# # 2. Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

In [29]:
df_ords.describe()

order_id       user_id  order_number    orders_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

Under count, all fields should have the same number.  However, days_since_prior_order is 3214874 instead of 3214873 like alll other fields.  Everything else looks normal.

# # 3. Check for mixed-type data in your df_ords dataframe.

In [32]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

No columns came back as having mixed-type data.

# #4. If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

I did not find any columns containing mixed-type data.

# #5. Run a check for missing values in your df_ords dataframe.

In [63]:
#Use isnull() function to check for missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_dow                     0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [65]:
df_ords.shape

(3421083, 7)

In [67]:
df_ords.head()

order_id  user_id eval_set  order_number  orders_dow  order_hour_of_day  \
0   2539329        1    prior             1           2                  8   
1   2398795        1    prior             2           3                  7   
2    473747        1    prior             3           3                 12   
3   2254736        1    prior             4           4                  7   
4    431534        1    prior             5           4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

There are 206,209 missing values for days_since_prior_order.  The days_since_prior_order shows how recently a customer has placed a previous order.  I would assume that the NaN responses represent customers who have never ordered from Instacart before, but have placed their first order.

# #6. Address the missing values using an appropriate method.

If I am correct and the NaN values represent first time customers, then I think the best way to handle the NaN is to leave them alone and use them as a flag for 'new_customer'.  Then Instacart could use this as an opportunity to offer coupons or special deals to first time customers in hopes of getting more repeat business from them.

# #7 & 8. Run a check for duplicate values in your df_ords data

In [74]:
df_dups = df_ords[df_ords.duplicated()]

In [76]:
df_dups

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, orders_dow, order_hour_of_day, days_since_prior_order]
Index: []

There is no duplicate data in the df_ords dataframe.

# #9. Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [90]:
#No changes were made, but I want to notate that this dataset is clean and ready to use.
df_ords_clean = df_ords

In [86]:
#Export df_ords_clean as csv
df_ords_clean.to_csv('/Users/ariedlin/Desktop/Career Foundry Assignments/8-24 Instacart Blanket Analysis Real/2 Data/2 Prepared Data/df_ords_clean.csv')